RavenDB-14635 Adding support for document / revisions compression

ayende opened the pull request:
Ready

ayende committed:
RavenDB-14635 Can compress table data using zstd with training dictionaries

ayende committed:
RavenDB-14635 Ensure that we properly check if a table entry is freed or not

ayende committed:
RavenDB-14635
- Revisions now use compression by default
- Be more conservative about creating new compression dictionaries

ayende committed:
RavenDB-14635
- Allowing to get the compression dictionary without using an active raw data section
- Using a single shared tree for all compression dictionaries

ayende commented:
Initial results, here is the sample data revisions without compression:

![image](https://user-images.githubusercontent.com/116915/76657678-12dbd100-657b-11ea-8fa4-f520201f92fe.png)

As you can see, we have 3.65 MB of table data. 

Here is the results with compression:

![image](https://user-images.githubusercontent.com/116915/76657714-2be48200-657b-11ea-9b98-c1cd98306fe7.png)

So total of 1.67 MB table data, so we are only taking 45% of the previous size.

As we have more revisions, we seems to get better rate, too.

ayende committed:
RavenDB-14635 Started to add the zstd lib to the project directly

ayende committed:
RavenDB-14635 Can define compressed collections in the db record

ayende committed:
RavenDB-14635 Handle move of data when we defrag pages of compressed records

ayende committed:
RavenDB-14635 Trying to handle decompressing of data while defragging a page

ayende committed:
RavenDB-14635 Better handling of document updates, handling of moving between compressed section when the document is not compressable to the next one

ayende committed:
RavenDB-14635 Avoid NRE when failing to create a new server

ayende committed:
RavenDB-14635 Better error messages, adding testing

ayende committed:
RavenDB-14635 Fixing native integration for multi assemblies

ayende committed:
RavenDB-14635 WIP

ayende committed:
RavenDB-14635 - Refactoring compression when allocating new segments

ayende committed:
RavenDB-14635 Fixing compression behavior on delete / defrag

ayende committed:
RavenDB-14635 Minor test fixes

ayende commented:
With compression:

![image](https://user-images.githubusercontent.com/116915/78501556-8d57d500-7764-11ea-9d9a-1d6fd7a3244c.png)


Without compression:

![image](https://user-images.githubusercontent.com/116915/78501565-98126a00-7764-11ea-869b-bb0e49baaec0.png)

We got it down by 20% !

ayende committed:
RavenDB-14635 - Now supporting adding compression to a collection after it has some documents.

ayende committed:
RavenDB-14846 Support compaction of databases with compressed collections

ayende commented:
This is a big change, and it isn't completely done. Still need to handle recovery here.
But it is ready for review.

ayende committed:
RavenDB-14756
- Use a numeric id and not the dictionary hash to identify dictionaries
- Allow to create sample data without indexes
- Don't compact the dictionaries tree, we'll regenerate it on compaction

topalach commented:
retest this please

topalach commented:
The previous commit was added to trigger jenkins PR tests.

ayende committed:
RavenDB-14756
- Use ~8KB for the size of the dictionary
- On update, dispose the decompressed value immediately
- Need to start dictionary id on 1, not 0 (to indicate missing compression on section)
- Better handling of corrupted data during recovery

ayende committed:
RavenDB-14756
- More robust blittable validation code
- Output dictionaries to a separate file to ease recovery
-

ayende committed:
RavenDB-14849 Recovery for compressed documents - should not use a single file for all dictionaries

ayende committed:
RavenDB-14849 Renaming a class to avoid naming collision

ayende committed:
RavenDB-14849 Avoid having to deal with size changes during insert / update process.
We now keep track of the dictionary per document individually and train documents based on the last documents changed (not per section)

ayende committed:
RavenDB-14849 Tuning compression performance / storage size

ayende committed:
RavenDB-14849 Refactoring doc compression configuration and adding licensing support

grisha-kotler commented on src/Raven.Server/Commercial/LicenseStatus.cs:
```diff
@@ -105,6 +105,11 @@ public int MaxClusterSize
 
         public bool HasEncryption => GetValue<bool>("encryption");
 
+        public bool HasCompression =>
```
We'll add the compression to all the licenses.
You also need to update `LicenseValidator`

ayende replied:
done

ayende committed:
RavenDB-14849 Recovery for compressed documents
Can now read the recovery dictionaries (stored in two separate files next to the data file).
Include recover dictionaries when creating snapshot backup
Fixing reading of big endian data from value reader

ayende committed:
RavenDB-14756 Account for the decompression buffer whenever we need to take into account the transaction memory costs

ayende committed:
RavenDB-14756 allow to disable revisions compression globally

ayende committed:
RavenDB-14756 Add compression to license

ayende committed:
RavenDB-14865 Wipe compression buffer after tx dispose

ayende committed:
RavenDB-14865 - Encrypt the compression dictionaries when the database is encrypted

ayende committed:
RavenDB-14865 Don't reset allocator that may be shared

ayende committed:
RavenDB-14635 Fix Linux build

ayende committed:
RavenDB-14635 Only enable compression feature in experimental mode

ayende committed:
RavenDB-14635 Only enable feature for revisions if we are running experimental features

arekpalinski commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -307,4 +309,48 @@ public enum DeletionInProgressStatus
+    public class CompressionConfiguration
+    {
+        public string[] Collections;
+        public bool CompressRevisions;
```
Given that we have an option to control compression of documents per collection then why not to let user control revisions per specific collection as well (with the global option for all revisions of course)?

ayende replied:
The default for revisions is that they are all included, because they are rarely touched.
Can you think of a reason why to have this for only specific collections?

arekpalinski replied:
Actually I was rather thinking about excluding some revisions from compression. For example a user might have heavy subscription on revisions from a particular collection, so faster access time and lower CPU usage can be preferred in that case.

ayende replied:
I think this is a rather niche scenario, and even so, we aren't likely to spend too much time of this. Subscriptions are no time sensitive, I think. 
The issue I have is that we would need `CompressRevisions` and `ExcludeRevisionCompression` and that seems like very rare scenario to need it.

arekpalinski replied:
I'm fine with that

arekpalinski commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -307,4 +309,48 @@ public enum DeletionInProgressStatus
+        protected bool Equals(CompressionConfiguration other)
+        {
+            return CompressRevisions == other.CompressRevisions && 
+                   Collections.SequenceEqual(other.Collections);
```
The order of elements shouldn't matter I think

arekpalinski commented on src/Raven.Server/Commercial/LicenseStatus.cs:
```diff
@@ -105,6 +105,11 @@ public int MaxClusterSize
+        public bool HasCompression =>
+            // TODO: we don't have compression in any license yet
+            GetValue<bool?>("compression") ??
+            GetValue<bool>("encryption");
```
Why do we default to "encryption"?

ayende replied:
There are a lot of licenses here without this flag, so we alias the encryption to allow this as well.
If the compression is set, it takes precedence.  So you can have a compression false, but with encryption

arekpalinski commented on src/Raven.Server/Documents/CompactDatabaseTask.cs:
```diff
@@ -103,7 +112,12 @@ public async Task Execute(Action<IOperationProgress> onProgress, CompactionResul
+                        }, (name, schema) =>
+                        {
+                            schema.Compressed = 
+                                name.StartsWith(revisionsPrefix,StringComparison.OrdinalIgnoreCase) ||
```
This sets `Compressed` to `true` for _any_ revision table - regardless `public bool CompressRevisions;` configuration option

arekpalinski commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -219,6 +221,36 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
+            {
+                return CompressedDocsSchema;
+            } 
+            return DocsSchema;
```
Given that the schema is different - do we need to prevent from changing the configuration once the compression was turned on for a given collection?

ayende replied:
What do you mean? The on disk schema is identical on both ends.

arekpalinski replied:
Okay, I see. Anyway, it is valid to turn off compression if it was already enabled?

ayende replied:
Yes

arekpalinski commented on src/Raven.Server/Raven.Server.csproj:
```diff
@@ -77,7 +77,25 @@
+    <!--
+    <None Include="..\..\libs\libzstd\libzstd.mac.x64.dylib" Link="libzstd.mac.x64.dylib">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>-->
```
Why commented?

ayende replied:
I don't have access to a mac to build and test it yet

ppekrol commented on src/Raven.Client/Exceptions/Commercial/LimitType.cs:
```diff
@@ -27,6 +27,9 @@ public enum LimitType
         Encryption,
+        
+        [Description("Document Compression")]
+        DocumentCompression,
```
What is the official name for this? Compression or Document Compression?

cc @Danielle9897

ayende replied:
I think we want to call it document compression

ppekrol commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1320,6 +1325,12 @@ private void ThrowIfCannotActivateLicense(LicenseStatus newLicenseStatus)
                 throw GenerateLicenseLimit(LimitType.DynamicNodeDistribution, message);
             }
+
+            if(!newLicenseStatus.HasCompression && documentCompressionCount > 0)
```
Let's stick either to Compression (too generic?) or Document Compression.

If we will go with 'DocumentCompression' then let's change this to `newLicenseStatus.HasDocumentCompression`

ppekrol commented on src/Raven.Server/Commercial/LicenseStatus.cs:
```diff
@@ -105,6 +105,11 @@ public int MaxClusterSize
 
+        public bool HasCompression =>
+            // TODO: we don't have compression in any license yet
+            GetValue<bool?>("compression") ??
```
"documentCompression" ?

ppekrol commented on src/Raven.Server/Commercial/LicenseValidator.cs:
```diff
@@ -16,7 +16,8 @@ public static class LicenseValidator
             "externalReplication", "delayedExternalReplication", "ravenEtl", "sqlEtl", "highlyAvailableTasks",
-            "pullReplicationAsHub", "pullReplicationAsSink", "encryptedBackup", "letsEncryptAutoRenewal", "cloud"
+            "pullReplicationAsHub", "pullReplicationAsSink", "encryptedBackup", "letsEncryptAutoRenewal", "cloud",
+            "compression"
```
"documentCompression" ?

ppekrol commented on src/Raven.Server/Config/Categories/DatabaseConfiguration.cs:
```diff
@@ -43,6 +43,16 @@ public class DatabaseConfiguration : ConfigurationCategory
+        [Description("Whatever revisions compression should be on by default or not on new databases")]
+        [DefaultValue(true)]
+        [ConfigurationEntry("Databases.DefaultRevisionsCompression", ConfigurationEntryScope.ServerWideOnly)]
+        public bool DefaultRevisionsCompression { get; set; }
```
I think this feature deserves dedicated configuration section. CompressionConfiguration with `Compression.ABC` keys.

ppekrol replied:
Maybe we should go with Compression name, similar to Revisions?

ppekrol commented on src/Raven.Server/Documents/DocumentPutAction.cs:
```diff
@@ -219,6 +221,36 @@ public DocumentPutAction(DocumentsStorage documentsStorage, DocumentDatabase doc
+            if (_compression.Equals(record.Compression))
+                return;
+
+            if (record.Compression == null) // legacy configurations
```
Will this be `null`? It is initialized in `DatabaseRecord` by default.

ayende replied:
Existing database records don't have this.

ppekrol commented on src/Raven.Server/Raven.Server.csproj:
```diff
@@ -77,7 +77,25 @@
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <!--
+    <None Include="..\..\libs\libzstd\libzstd.mac.x64.dylib" Link="libzstd.mac.x64.dylib">
```
Do we have a issue for this?

ayende replied:
https://issues.hibernatingrhinos.com/issue/RavenDB-14840

ppekrol commented on src/Raven.Server/Storage/Schema/SchemaUpgrader.cs:
```diff
@@ -16,7 +16,7 @@ internal class CurrentVersion
             public const int ConfigurationVersion = 11;
 
-            public const int DocumentsVersion = 17;
+            public const int DocumentsVersion = 18;
```
We need to apply https://github.com/ravendb/ravendb/pull/9300

ppekrol replied:
Can you rebase? It should be fixed now.

ppekrol commented on src/Raven.Server/Utils/DebuggerAttachedTimeout.cs:
```diff
@@ -5,7 +5,7 @@ namespace Raven.Server.Utils
     public static class DebuggerAttachedTimeout
     {
-        public static bool DisableLongTimespan;
+        public static bool DisableLongTimespan = true;
```
Remove?

ayende replied:
I think we want to keep it.
We need it enabled only in rare cases, but it hurts usability on debugging

ppekrol commented on src/Raven.Server/Web/Studio/SampleDataHandler.cs:
```diff
@@ -23,6 +23,7 @@ public async Task PostCreateSampleData()
         {
             using (ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
             {
+                var skipIndexes = GetBoolValueQueryString("skipIndexes", false) ?? false;
```
Why this is needed?

ayende replied:
I want to be able to create sample database without creating indexes.
Useful for tests where we test compression only

ppekrol replied:
Do we want to expose this in the UI?

ayende replied:
I don't think so, it is very rarely going to be useful for users.

ppekrol commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -132,6 +134,12 @@ public async Task AddDatabaseNode()
                 if (databaseRecord.Encrypted)
                     ServerStore.LicenseManager.AssertCanCreateEncryptedDatabase();
 
+                if (databaseRecord.Compression?.CompressRevisions == true ||
```
Aren't we setting this value in ctor by default to true?

ayende replied:
When reading from old records, it it still null

ppekrol commented on src/Raven.Server/Web/System/AdminDatabasesHandler.cs:
```diff
@@ -241,9 +249,27 @@ public async Task Put()
+                {
+                    ServerStore.LicenseManager.AssertCanUseDocumentCompression();
+                    if (Server.Configuration.Core.FeaturesAvailability != FeaturesAvailability.Experimental)
+                        FeaturesAvailabilityException.Throw("Compression");
```
Compression vs Document Compression

ppekrol commented on src/Voron/Data/Tables/Table.cs:
```diff
@@ -174,21 +193,72 @@ private bool TryFindIdFromPrimaryKey(Slice key, out long id)
+                return t.Ptr;
+            }
+
+            //TODO: For encrypted databases, we need to allocate on locked memory
```
TODO. Do we have an issue for this?

ayende replied:
No, we need to remove it. This is handled by wiping the allocator memory

ppekrol commented on src/Voron/Data/Tables/Table.cs:
```diff
@@ -333,13 +417,25 @@ public void Delete(long id)
-            var ptr = DirectRead(id, out int size);
-            if (ptr == null)
-                return;
+            var ptr = DirectReadRaw(id, out int size, out bool compressed);
```
Can this be null?

ayende replied:
No, this must return a value.

arekpalinski commented on test/FastTests/Server/Documents/Indexing/Static/BasicStaticMapIndexing.cs:
```diff
@@ -500,10 +500,10 @@ public void NumberOfDocumentsAndTombstonesToProcessShouldBeCalculatedCorrectly()
+      //  [InlineData(1000, 2000)]
         [InlineData(128, 2048)]
-        [InlineData(null, 1000)]
+    //    [InlineData(null, 1000)]
```
Why is it commented?

ayende replied:
Should restore it

ppekrol commented on src/Voron/Data/Tables/Table.cs:
```diff
@@ -704,59 +921,56 @@ internal FixedSizeTree GetFixedSizeTree(Tree parent, Slice name, ushort valSize,
+        private void CreateNewActiveSection()
         {
-            if (ActiveDataSmallSection.TryAllocate(size, out long id) == false)
+            ushort maxSectionSizeInPages =
```
Something that came to my mind. We need to check if moving between 32 and 64 bits work.

ayende replied:
It *shouldn't* matter, but we are taking more memory in this case, I think

arekpalinski replied:
This look _very_ familiar to the issue we had with `NewPageAllocator` and we ended up with persisting the actual allocation size - see here:

https://github.com/ravendb/ravendb/pull/9610

ayende replied:
The number of pages is persisted, see inside the ActiveRawDataSection

ayende replied:
Note that this part is just moved code, I didn't change functionality from prev behavior

arekpalinski reviewed (commented):
This is not my final review. I'll go over the compression code once again.

ppekrol commented on src/Voron/Data/Tables/TableSchema.cs:
```diff
@@ -27,8 +27,13 @@ public unsafe class TableSchema
         public static readonly Slice StatsSlice;
         public static readonly Slice SchemasSlice;
         public static readonly Slice PkSlice;
+        public static readonly Slice DictionariesSlice;
```
CompressionDictionariesSlice

ppekrol commented on src/Voron/Data/Tables/TableSchema.cs:
```diff
@@ -27,8 +27,13 @@ public unsafe class TableSchema
         public static readonly Slice SchemasSlice;
         public static readonly Slice PkSlice;
+        public static readonly Slice DictionariesSlice;
+        public static readonly Slice CompressionDictionaryIdSlice;
```
CurrentCompressionDictionaryIdSlice?

ppekrol commented on src/Voron/Data/Tables/TableValueCompressor.cs:
```diff
@@ -0,0 +1,401 @@
+            // the idea is that we'll get better results by including the most recently modified documents
+            // by iterating over the tag index, which is guaranteed to be always increasing
+            var dataIds = new List<long>(256);
+            var sizes = new UIntPtr[256];
```
Should we use ArrayPool here?

ppekrol commented on src/Voron/Impl/Backup/FullBackup.cs:
```diff
@@ -167,6 +167,32 @@ public class StorageEnvironmentInformation
                 {
+                    if (env.Options is StorageEnvironmentOptions.DirectoryStorageEnvironmentOptions)
+                    {
+                        foreach (var recoveryFile in Directory.GetFiles(env.Options.BasePath.FullPath, "*.Recovery"))
```
Why are we taking those files now?

ayende replied:
So recovery of database snapshot could work

ppekrol commented on src/Voron/ValueReader.cs:
```diff
@@ -79,9 +79,6 @@ public int ReadLittleEndianInt32()
 
             _pos += sizeof (int);
 
-            if (!BitConverter.IsLittleEndian)
```
Why it was removed?

ayende replied:
Gave wrong results

arekpalinski commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -122,25 +128,10 @@ full name
-            });
+            DefineIndexesForDocsSchema(DocsSchema);
+            DefineIndexesForDocsSchema(CompressedDocsSchema);
+            DocsSchema.CompressValues(CompressedDocsSchema.FixedSizeIndexes[CollectionEtagsSlice], compress: false);
```
```suggestion
            DocsSchema.CompressValues(DocsSchema.FixedSizeIndexes[CollectionEtagsSlice], compress: false);
```

Minor but shouldn't be this way?

arekpalinski commented on src/Voron/Data/RawData/ActiveRawDataSmallSection.cs:
```diff
@@ -22,9 +22,22 @@ namespace Voron.Data.RawData
+            _transaction = tx;
+        }
+
+        public static int CompressionDictionaryHashFor(LowLevelTransaction tx,long id)
```
Where do we use it? (I can't find any usage on github)

ayende replied:
dead code

arekpalinski commented on src/Voron/Data/Tables/Table.cs:
```diff
@@ -174,21 +193,72 @@ private bool TryFindIdFromPrimaryKey(Slice key, out long id)
+            var internalScope = tx.Allocator.Allocate(decompressedSize, out buffer);
+            var actualSize = ZstdLib.Decompress(data, buffer.ToSpan(), dictionary);
+            if (actualSize != decompressedSize)
+                throw new InvalidDataException($"Got decompressed size {actualSize} but expected {decompressedSize}");
```
Can we add more details here?
- transaction id
- dictionary id
- can we include the entire dictionary? would it make sense?

I'm trying to collect as much details as possible in case we'll need that for troubleshooting

ayende replied:
The actual dictionary is 8KB in size, and is binary.
Doesn't make sense.

It is there is we messed up the numbers, mostly.

arekpalinski replied:
Okay, but transaction id and dictionary id might be helpful when investigating some corruption issues

ayende replied:
Done

arekpalinski commented on src/Voron/Data/RawData/RawDataSection.cs:
```diff
@@ -24,16 +27,47 @@ public unsafe class RawDataSection : IDisposable
+                    Debug.Assert(IsFreed == false, "Should not try to get size of freed instance");
+                    return (short)(UsedSize_Buffer & ValueOnlyMask);
+                }
+                set => UsedSize_Buffer = (short)(value & ValueOnlyMask);
```
It's meant to be called only for non-compressed data, right?

ayende replied:
Yes, this is called, and then the is compressed is set.

arekpalinski commented on src/Voron/Data/Tables/Table.cs:
```diff
@@ -174,21 +193,72 @@ private bool TryFindIdFromPrimaryKey(Slice key, out long id)
+
+            //TODO: For encrypted databases, we need to allocate on locked memory
+
+            // we explicitly do *not* dispose the buffer, it lives as long as the tx
```
This means that for long running transactions we might have a lot of cached values even if we don't need them anymore. Maybe we should introduce LRU cache and clear old values if the cache grows too much?

ayende replied:
We can't, someone may be looking at this value.
We are accounting for that in the tx sizes.

arekpalinski replied:
Can we have the concept of tracking the actual usage and count it?

ayende replied:
Yes, this is something that we have to do. 
https://issues.hibernatingrhinos.com/issue/RavenDB-14870

arekpalinski commented on src/Voron/Data/Tables/Table.cs:
```diff
@@ -389,15 +490,32 @@ public void Delete(long id)
+
+                OnDataMoved(idToMove, newId, dataPtr, dataSize);
+
+                decompressedScope.Dispose();
```
Shouldn't the dispose be in inside `finally`?

ayende replied:
Same, this is perf sensitive, and I want to avoid the try/finally costs.
It isn't an issue if we leak, because the tx will dispose the values anway

arekpalinski replied:
OK. Worth to comments that this is intentional because of performance

ayende replied:
done

arekpalinski commented on src/Voron/Data/Tables/Table.cs:
```diff
@@ -333,13 +417,25 @@ public void Delete(long id)
             var tvr = new TableValueReader(ptr, size);
             DeleteValueFromIndex(id, ref tvr);
 
+            decompressValue.Dispose();
```
Shouldn't be inside `finally` or just wrap by `using`?

ayende replied:
I want to avoid either options for perf costs. 
If we leak, this will be disposed by the tx closing anyway

arekpalinski commented on src/Voron/Data/Tables/Table.cs:
```diff
@@ -518,7 +629,102 @@ public long Insert(TableValueBuilder builder)
+        private void ThrowBadWriter(int size)
+        {
+            throw new VoronErrorException(
+                $"After successfully allocating {size:#,#;;0} bytes, failed to write them on {Name}");
```
Can we add more details here? Basically all parameters that we pass to `TryWriteDirect`:
- id
- size
- compressed

arekpalinski commented on src/Voron/Data/Tables/Table.cs:
```diff
@@ -704,59 +921,56 @@ internal FixedSizeTree GetFixedSizeTree(Tree parent, Slice name, ushort valSize,
+                            ActiveCandidateSection.Delete(sectionPageNumber);
+                            {
+                                return true;
+                            }
```
Wrapped by `{ }` ?

arekpalinski commented on src/Voron/Data/Tables/TableValueBuilder.cs:
```diff
@@ -175,5 +199,41 @@ public void CopyTo(byte* ptr)
+            CopyTo(Compression.RawBuffer.Ptr);
+
+            return Compression.TryCompression(table, schema);
+        }
```
Do/should we return memory allocated here of `TryCompression` returns `false`?

ayende replied:
No, this is going to be cleanup anyway shortly when the operation is done.

arekpalinski commented on src/Voron/Data/Tables/TableValueCompressor.cs:
```diff
@@ -0,0 +1,401 @@
+        {
+            var count = 0;
+            var v = (uint)value;
+            while (v >= 0x80)
```
Can we create const for `0x80`?

ayende replied:
No, this is actually clearer when you see the constant.
Because it follows the structure of varint.

arekpalinski commented on src/Voron/Data/Tables/TableValueCompressor.cs:
```diff
@@ -0,0 +1,401 @@
+            while (v >= 0x80)
+            {
+                buffer[count++] = (byte)(v | 0x80);
+                v >>= 7;
```
Let's add comment explaining `>>= 7` ?

ayende replied:
That is explained elsewhere in the code, this is the same varint we use everywhere

arekpalinski commented on src/Voron/Data/Tables/TableValueCompressor.cs:
```diff
@@ -0,0 +1,401 @@
+                    sizes[dataIds.Count] = (UIntPtr)size;
+                    dataIds.Add(id);
+                    totalSize += size;
+                } while (dataIds.Count < 256 && it.MovePrev() && totalSize < 1024 * 1024);
```
What is the reasoning that we iterate from the last element?

ayende replied:
We want to get the most recently update values, we assume that those are the most similar to the currently inserted dta.

arekpalinski commented on src/Voron/Data/Tables/TableValueCompressor.cs:
```diff
@@ -0,0 +1,401 @@
+
+            var rev = Bits.SwapBytes(newId);
+            using var _____ = Slice.External(tx.Allocator, (byte*)&rev, sizeof(int), out var slice);
+            using var ____ = dictionariesTree.DirectAdd(slice, sizeof(CompressionDictionaryInfo) + dictionaryBufferSpan.Length, out var dest);
```
This is very crucial piece of code - I'd prefer to see explicitly the scopes of `usings` in this method. It's easier to read it IMO.

arekpalinski commented on src/Voron/Data/Tables/TableValueCompressor.cs:
```diff
@@ -0,0 +1,401 @@
+                var dictionaries = tx.ReadTree(TableSchema.DictionariesSlice);
+
+                if (dictionaries == null)
+                    return; // should never happen
```
Debug.Assert ?

arekpalinski commented on src/Voron/Data/Tables/TableValueCompressor.cs:
```diff
@@ -0,0 +1,401 @@
+                        .Combine(path: $"Compression{(i == 0 ? "A" : "B")}.Recovery")
+                        .FullPath;
+
+                    using var fs = File.Open(newPath, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.Read);
```
Don't we need to call `fs.Flush(flushToDisk: true);` at the end?

ayende replied:
I intentionally didn't do that. 
We have two separate files here, and I want to avoid flushing the entire fs cache
That can kill our perf

arekpalinski replied:
1. But if non of the files is flushed then how can we be guaranteed that at least one have everything synced? 
2. This is called once per write transaction, right?

ayende replied:
This is called per value saved.

ayende replied:
Yes, sorry, once per write tx that created new dictionary
And we don't flush because:
* We have the data saved in the data file
* We re-create it on new dictionarys

arekpalinski commented on src/Voron/Data/Tables/ZstdLib.cs:
```diff
@@ -0,0 +1,279 @@
+            }
+
+            ~CompressionDictionary()
+            {
```
Maybe let's add some logging or just console write line in DEBUG when the finalizer is called

ayende committed:
RavenDB-14635 Fixing PR comments

ppekrol commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -54,6 +54,8 @@ public DatabaseRecord(string databaseName)
 
         public ConflictSolver ConflictSolverConfig;
 
+        public CompressionConfiguration Compression = new CompressionConfiguration(compressRevisions: true);
```
Compression => DocumentCompression

ayende replied:
Done

ppekrol commented on src/Raven.Client/ServerWide/DatabaseRecord.cs:
```diff
@@ -307,4 +309,50 @@ public enum DeletionInProgressStatus
         HardDelete
     }
+
+    public class CompressionConfiguration
```
DocumentCompressionConfiguration

ayende replied:
Done

ppekrol commented on src/Raven.Server/Commercial/LicenseStatus.cs:
```diff
@@ -105,6 +105,11 @@ public int MaxClusterSize
 
+        public bool HasDocumentCompression =>
+            // TODO: we don't have compression in any license yet
+            GetValue<bool?>("documentsCompression") ??
```
documentCompression

ayende replied:
Why singular here?

ppekrol commented on src/Raven.Server/Commercial/LicenseValidator.cs:
```diff
@@ -16,7 +16,8 @@ public static class LicenseValidator
             "externalReplication", "delayedExternalReplication", "ravenEtl", "sqlEtl", "highlyAvailableTasks",
-            "pullReplicationAsHub", "pullReplicationAsSink", "encryptedBackup", "letsEncryptAutoRenewal", "cloud"
+            "pullReplicationAsHub", "pullReplicationAsSink", "encryptedBackup", "letsEncryptAutoRenewal", "cloud",
+            "documentsCompression"
```
documentCompression

ppekrol commented on src/Raven.Server/Config/Categories/DatabaseConfiguration.cs:
```diff
@@ -43,6 +43,16 @@ public class DatabaseConfiguration : ConfigurationCategory
+        [Description("Whatever revisions compression should be on by default or not on new databases")]
+        [DefaultValue(true)]
+        [ConfigurationEntry("Databases.Compression.CompressRevisionsDefault", ConfigurationEntryScope.ServerWideOnly)]
+        public bool CompressRevisionsDefault { get; set; }
```
Wouldn't it be better to move this to DocumentCompressionConfiguration?

ayende replied:
We have just a single value, and it is already possible to change the code structure without the user seeing it. I don't think so.

ppekrol commented on src/Raven.Server/Storage/Schema/Updates/Documents/From17.cs:
```diff
@@ -0,0 +1,14 @@
+﻿namespace Raven.Server.Storage.Schema.Updates.Documents
+{
+    public class From17 : ISchemaUpdate
```
This can be removed.

ayende replied:
done

ppekrol commented on src/Voron/Slice.cs:
```diff
@@ -234,6 +234,13 @@ public Span<byte> AsSpan()
+        {
+            if (a.Size != b.Size)
+                return false;
+            return Memory.Compare(a.Content.Ptr, b.Content.Ptr, a.Size) == 0;
```
Shouldn't we use SliceComparer here?

ayende replied:
Removed it entirely.

ppekrol commented on src/Voron/Data/Tables/TableValueCompressor.cs:
```diff
@@ -0,0 +1,419 @@
+                for (int i = 0; i < 2; i++)
+                {
+                    var newPath = obj.Environment.Options.BasePath
+                        .Combine(path: $"Compression{(i == 0 ? "A" : "B")}.Recovery")
```
Can we move `.Recovery` to some const and use it here and in the Recovery.cs ? Maybe also consider changing the extension to `.compression-recovery`? It is easier to recognize the files via extension than the name. Some day we might add more 'recovery' files.

ayende replied:
done

ayende committed:
RavenDB-14635 Handle compression recovery after rebase

ayende committed:
RavenDB-14635 PR comments fixes

ayende committed:
RavenDB-14635 Avoid reading pages that we haven't checked

ayende committed:
RavenDB-14635 Retrain dictionary on the 1024th document that doesn't compress well enough.
This ends up being very effective because we aren't hitting this very often, but when we need to retrain, we getting there fast.

ayende committed:
RavenDB-14635 We can't set DocumentsCompression by default. If users don't enable experimental features, they will fail to create the database.
Instead, we handle that at the database creation stage

ayende committed:
RavenDB-14635 Fix test

ayende committed:
RavenDB-14635 "Document Compression" to "Documents Compression"

ppekrol commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1321,6 +1326,12 @@ private void ThrowIfCannotActivateLicense(LicenseStatus newLicenseStatus)
+
+            if(!newLicenseStatus.HasDocumentsCompression && documentCompressionCount > 0)
+            {
+                var message = GenerateDetails(documentCompressionCount, "document compression");
```
"documents compression"

ppekrol commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1441,6 +1452,19 @@ public void AssertCanAddExternalReplication()
+            if (_licenseStatus.HasDocumentsCompression)
+                return;
+
+            var details = $"Your current license ({_licenseStatus.Type}) does not allow document compression";
```
documents compression

ayende committed:
RavenDB-14635 More "Document Compression" to "Documents Compression"

arekpalinski reviewed (approved):
LGTM. Please just rebase to use zsdt libs build by @gregolsky. Also after rebase we should see multiplatform FastTest results so it will test if zsdt dlls are compiled correctly.

ppekrol committed:
RavenDB-14635 fixing csproj

ppekrol committed:
RavenDB-14635 compression tests require a valid license

ppekrol committed:
RavenDB-14635 disabled documents compression for RavenDB_14881 test
